aws/aws-sdk-go-v2
# Add ApiGatewayExportsNullabilityExceptionIntegration

For changes to files under the `/codegen/aws-models` folder, and manual edits to autogenerated code (e.g. `/service/s3/api.go`) please create an Issue instead of a PR for those type of changes.

N/A.

---

If the PR addresses an existing bug or feature, please reference it here.

N/A.

---

To help speed up the process and reduce the time to merge please ensure that `Allow edits by maintainers` is checked before submitting your PR. This will allow the project maintainers to make minor adjustments or improvements to the submitted PR, allow us to reduce the roundtrip time for merging your request.

---

## Overview

Due to internal model fixes for API Gateway (APIGW) exports, certain shapes
and members that target those shapes that used to have defaults became
nullable. In order to not break the existing Go v2 client API nullability,
this customization adds back default values for affected shapes:

- Root Boolean shapes: false
- Root Number shapes: 0
- Snapshotted Members shapes: inherit defaults

The class of services affected are APIGW services and APIGW services that
migrated to Smithy, seen in `APIGW_NULLABILITY_EXCEPTION_SERVICES`.

A "snapshot" of root-level and member shapes are captured in
`APIGW_exports_nullability_exceptions.json`, a mapping of service shape IDs
to a list of affected root-level and member shapes.

Cases:
- Snapshotted Root level:
  - If a snapshotted root level shape doesn't exist, an exception will be
    thrown.
  - Otherwise, a default will be patched if a default doesn't already exist
- Snapshotted Member level:
  - If a snapshotted member level shape doesn't exist, an exception will be
    thrown.
  - Otherwise, a default will be patched if a default doesn't already exist
- Nonsnapshotted Member level:
  - All nonsnapshotted member level shape that target a snapshotted root level
    will be identified and throw an error.
  - This is prevent breaking changes if a nonsnapshotted member changes target
    from a snapshotted root shape to a nonsnapshotted root shape, as the SDK
    has no prior info on either shape.

Also removes `serviceid.smithy` which was not used anywhere.

## Testing

Surprisingly, Go v2 had NO codegen tests at all. This PR sets up the first codegen test.

1. CI passes
2. Add `ApiGatewayExportsNullabilityExceptionIntegrationTest`
3. No diffs for code generation with existing service models when running `make generate`
4. No diffs for code generation when stripping `@default` traits from the affected service models:
    - Strip `@default` traits from affected service models using the `crusher.js` script
    - Run `make generate` and confirm no codegen changes (empty commit)
    - Revert stripping `@default` traits
